intel-sgx/scripts/gdb.py

```diff
@@ -30,8 +38,35 @@ def read_long(inferior, address, length):
     ret += b
   return ret
 
-def find_vma_base(addr):
-  s = gdb.execute("info proc mappings", False, True)
+def find_vma_base_from_runner(addr):
+  """Given an address `addr`, find the baseaddress of the enclave that contains the `addr`.
+
+  This function uses a feature of the new ftxsgx-runners (2023-08),
+  and works on both the Montgomery loader and the Augusta loader.
+  """
+  lang_restore = gdb.parameter('language')
+  # Old versions of GDB such as gdb-9.2 on Ubuntu-20.04 do not have `gdb.with_parameter`.
+  # We therefore have to override the language first,
+  # and then carefully restore the old value after.
+  gdb.execute("set language c", from_tty = False, to_string = True)
+
+  # Example output: "$3 = 140737219919872"
+  func_result = gdb.execute(
+    "print (unsigned long) ftxsgx_get_baseaddress_from_pointer({})".format(addr),
+    from_tty = False,
+    to_string = True
+  )
+  func_result = int(func_result.strip().rpartition(" ")[-1])
+  gdb.execute("set language {}".format(lang_restore), from_tty = False, to_string = True)
+  return func_result
+
+def find_vma_base_from_proc_mappings(addr):
+  """Given an address `addr`, find the baseaddress of the enclave that contains the `addr`.
+
+  This function uses `info proc mappings` to do so, thus working on pre-2023-08 ftxsgx-runners,
+  but only works on the Montgomery loader (does not work on the Augusta loader).
+  """
+  s = gdb.execute("info proc mappings", from_tty = False, to_string = True)
```

Comment:
What happens when we use an old runner with an augusta loader? Is there a way to detect that this is not working as expected and returning a proper error message?

Reply:
I guess it should be technically possible. I wouldn't be confident with releasing code that I've never tested though, so this will have to happen on a virtual machine that supports this on a hardware level.

My thinking is to raise a follow-up issue and continue with this one without the extra logic. Thoughts?

Reply:
The [original issue](https://github.com/fortanix/rust-sgx/issues/301#issuecomment-964343555) shows that it likely results in `sgxstate auto` throwing a Python exception. So a follow-up issue may be appropriate. We currently also don't have a good way to track all different versions at runtime and tying them together.